src/font/face/coretext.zig

```diff
@@ -637,96 +627,114 @@ pub const Face = struct {
         };
 
         // Read the 'hhea' table out of the font data.
-        const hhea: opentype.Hhea = hhea: {
+        const hhea_: ?opentype.Hhea = hhea: {
             const tag = macos.text.FontTableTag.init("hhea");
-            const data = ct_font.copyTable(tag) orelse return error.CopyTableError;
+            const data = ct_font.copyTable(tag) orelse break :hhea null;
             defer data.release();
             const ptr = data.getPointer();
             const len = data.getLength();
             break :hhea opentype.Hhea.init(ptr[0..len]) catch |err| {
-                return switch (err) {
-                    error.EndOfStream => error.InvalidHheaTable,
-                };
+                log.warn("error parsing hhea table: {}", .{err});
+                break :hhea null;
             };
         };
 
-        const units_per_em: f64 = @floatFromInt(head.unitsPerEm);
+        const units_per_em: f64 =
+            if (head_) |head|
+                @floatFromInt(head.unitsPerEm)
+            else
+                @floatFromInt(self.font.getUnitsPerEm());
         const px_per_em: f64 = ct_font.getSize();
         const px_per_unit: f64 = px_per_em / units_per_em;
 
         const ascent: f64, const descent: f64, const line_gap: f64 = vertical_metrics: {
-            const hhea_ascent: f64 = @floatFromInt(hhea.ascender);
-            const hhea_descent: f64 = @floatFromInt(hhea.descender);
-            const hhea_line_gap: f64 = @floatFromInt(hhea.lineGap);
-
-            if (os2_) |os2| {
-                const os2_ascent: f64 = @floatFromInt(os2.sTypoAscender);
-                const os2_descent: f64 = @floatFromInt(os2.sTypoDescender);
-                const os2_line_gap: f64 = @floatFromInt(os2.sTypoLineGap);
-
-                // If the font says to use typo metrics, trust it.
-                if (os2.fsSelection.use_typo_metrics) break :vertical_metrics .{
-                    os2_ascent * px_per_unit,
-                    os2_descent * px_per_unit,
-                    os2_line_gap * px_per_unit,
-                };
-
-                // Otherwise we prefer the height metrics from 'hhea' if they
-                // are available, or else OS/2 sTypo* metrics, and if all else
-                // fails then we use OS/2 usWin* metrics.
-                //
-                // This is not "standard" behavior, but it's our best bet to
-                // account for fonts being... just weird. It's pretty much what
-                // FreeType does to get its generic ascent and descent metrics.
+            if (hhea_) |hhea| {
```

Comment:
I'd flip this into a `const hhea = hhea_ orelse` block. I think you'd lower a lot of nesting that way since it seems the else case is a lot shorter than the if case.